src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,191 @@
+use super::{BTreeMap, Felt, Rpo256, RpoDigest, StarkField, Vec, Word, WORD_SIZE};
+
+mod bits;
+pub use bits::BitsIterator;
+
+mod content;
+pub use content::{Content, ContentType};
+
+mod index;
+pub use index::NodeIndex;
+
+mod storage;
+pub use storage::{Storage, StorageError};
+
+#[cfg(test)]
+mod tests;
+
+// TIERED SPARSE MERKLE TREE
+// ================================================================================================
+
+pub struct TieredSmt {
+    // TODO this should be a constant
+    empty_subtrees: Vec<Word>,
+    storage: Storage,
+}
+
+impl TieredSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    pub const MAX_DEPTH: usize = 64;
+    pub const TIER_DEPTH: usize = 4;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn with_storage(storage: Storage) -> Self {
+        // Construct empty node digests for each layer of the tree
+        let empty_subtrees = (0..Self::MAX_DEPTH + 1)
+            .scan(RpoDigest::default(), |state, _| {
+                let value = *state;
+                *state = Rpo256::merge(&[value, value]);
+                Some(value.into())
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+            .collect();
+
+        Self {
+            empty_subtrees,
+            storage,
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn peek_node_type(&self, index: &NodeIndex) -> Result<ContentType, StorageError> {
+        self.storage
+            .peek_node_type(index)
+            .map(Option::unwrap_or_default)
+    }
+
+    pub fn get_node_key(&self, index: &NodeIndex) -> Result<Word, StorageError> {
+        self.storage
+            .get_node_key(index)?
+            .or_else(|| self.empty_subtrees.get(index.depth() as usize).copied())
+            .ok_or(StorageError::InvalidNodeIndex)
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn insert(&self, key: Word, value: Word) -> Result<(), StorageError> {
+        self.insert_key(key)?;
+        self.storage.insert_value(key, value)?;
+        Ok(())
```

Comment:
I'm not sure I fully understand how this would work. It seems like that we re-compute the root of the tree in `self.insert_key()` and this does not depend on the value being stored (since `insert_key()` does not take the value as the parameter). But the root of the tree should depend on the value as well (i.e., `insert(k0, a)` and `insert(k0, b)` should result in different tree roots.

Am I missing something?

Reply:
Separately, we should probably add public accessor to return the root of the tree.

Reply:
Indeed this is an open question. The point I need some clarification is that the key of a leaf is computed as `hash(path + depth ⋅ 2^48, v)`, so it means it can't be an arbitrary argument of `insert` (i.e. the user cannot define his key, the state of the tree + the value will define it). Maybe I'm missing something? We could have a quick sync call to clarify this important point. cc @Al-Kindi-0

Reply:
So, the rule for computing note hashes is as follows:

* Internal node: `hash(left, right, domain = 0)`
* Leaf at depth 16, 32, 48: `hash(remaining_path, value, domain = depth)`
* Leaf at depth 64: `hash(remaining_path, values_hash, domain = 64)`, where `values_hash` is a sequential hash of all key-value pairs in the subtree defined by this node. The key-value pairs are `(remaining_path, value)` and are sorted by `remaining_path`.

Reply:
> it means it can't be an arbitrary argument of `insert` (i.e. the user cannot define his key, the state of the tree + the value will define the node in which the value will be stored, and the key will be constant regardless of the depth).

The user can define an arbitrary key, but where the value under that key is stored in the tree depends on what else is already stored in the tree.

Reply:
Gotcha! I had a different initial understanding. I'll update the code and later today should be ready for a new round, thanks!

Reply:
Here is another comment that actually implies the tree key is `hash(user_key)` [ref](https://github.com/0xPolygonMiden/crypto/issues/22#issuecomment-1367559655)

> Also, it is possible to intentionally come up with a collision doing just work, but finding a 3rd, 4th etc.

So to me it seems we are missing the key in the hash. Shouldn't the hashes be this instead?

- Internal node: `hash(left, right, domain = 0)`
- Leaf at depth 16, 32, 48: `hash(hash(user_key)[used:], value, domain = depth)`
- Leaf at depth 64: `hash(values_hash, domain = 64)`, where `values_hash` is a `hash([(hash(user_key), value for user_key, value in all_entries])`

And in that case the sorting of the entries at depth 64 would be much more complicated, because it would need to sort the user's key, and they can be of arbitrary size. Am I missing something?

Reply:
> Question: At first I thought we would do `hash(user_key)` and use that value as the key for the Merkle tree. But from this comment, it seems you're saying we will use `key`. Is this correct? Wouldn't that make it trivial for collisions to happen? Or am I missing something?

The SMT does not by itself dictate the structure of keys and values - but how it is used in different places may impose this structure. For example, when the SMT is used for account storage, the key will be hashed first first, and then the result will be used as a key for the SMT. However, when we use SMT for account vault, we don't need to do this hashing as the structure of assets makes collisions unlikely.

> Another question, why are we changing the domain, i.e. `hash(remaining_path, value, domain = depth)`, instead of doing `hash(remaining_path, value, depth)` (as in, adding the depth as part of the data being hashed)

This is for efficiency reasons. `hash(remaining_path, value, depth)` would require 2 permutations of the hash function. By using domains we can do it in a single permutation but we do give up 2 bits of security. I don't think that it's a big deal though.

> And in that case the sorting of the entries at depth 64 would be much more complicated, because it would need to sort the user's key, and they can be of arbitrary size. Am I missing something?

The key is always 4 elements and by the time we get to depth 64, one of these elements is already "accounted for". So, the remaining key is exactly 3 field elements. It will take some number of cycles to sort by these keys, but it shouldn't be too bad and also the likelihood that we'll and up with a lot of collisions at depth 64 is rather small.

Reply:
Thanks for the reply @bobbinth , I wish I could have done my updates a bit faster to save you a bit of time 😄 

> The SMT does not by itself dictate the structure of keys and values - but how it is used in different places may impose this structure [...] However, when we use SMT for account vault, we don't need to do this hashing as the structure of assets makes collisions unlikely.

Ah, I haven't thought about using an asset as key, makes sense to avoid re-hashing it, cool!

Just to clarify, AFAIU we do impose structure in the keys, namely that the key must be 4 Felts in length and it must be randomly distributed. Asset happens to fit that description, or am I missing something?

It would be interesting to represent that in the type system, with a trait, perhaps something like `trait IntoDigest`. Structs like `asset` which already fit the bill would just return self, other structs would hash themselves first.

> This is for efficiency reasons. hash(remaining_path, value, depth) would require 2 permutations of the hash function.

Aha! clever, thanks for the explanation.